src/herder/TxSetFrame.cpp

```diff
@@ -797,57 +779,35 @@ TxSetFrame::addTxsFromXdr(Hash const& networkID,
 }
 
 void
-TxSetFrame::surgePricingFilter(uint32_t opsLeft)
+TxSetFrame::applySurgePricing(Application& app)
 {
     ZoneScoped;
-    auto curSizeOps = sizeOp();
-    if (curSizeOps <= opsLeft)
+
+    if (mTxs.empty())
     {
+        mFeesComputed = true;
         return;
     }
-    CLOG_WARNING(Herder, "surge pricing in effect! {} > {}", curSizeOps,
-                 opsLeft);
 
-    auto actTxQueueMap = TxSetUtils::buildAccountTxQueues(mTxs);
+    size_t maxOps = app.getLedgerManager().getLastMaxTxSetSizeOps();
+    auto const& lclHeader =
+        app.getLedgerManager().getLastClosedLedgerHeader().header;
 
-    std::priority_queue<TxSetFrame::AccountTransactionQueue*,
-                        std::vector<TxSetFrame::AccountTransactionQueue*>,
-                        SurgeCompare>
-        surgeQueue;
+    auto actTxQueues = TxSetUtils::buildAccountTxQueues(mTxs);
 
-    for (auto& am : actTxQueueMap)
-    {
-        surgeQueue.push(&am.second);
-    }
+    auto includedTxs = SurgePricingPriorityQueue::getMostTopTxsWithinLimit(
+        std::vector<TxStackPtr>(actTxQueues.begin(), actTxQueues.end()),
+        maxOps);
 
-    TxSetFrame::Transactions filteredTxs;
-    filteredTxs.reserve(sizeTx());
-    while (opsLeft > 0 && !surgeQueue.empty())
+    int64_t lowestFee = std::numeric_limits<int64_t>::max();
+    for (auto const& tx : includedTxs)
     {
-        auto cur = surgeQueue.top();
-        surgeQueue.pop();
-        // inspect the top candidate queue
-        auto& curTopTx = cur->front();
-        auto opsCount = curTopTx->getNumOperations();
-        if (opsCount <= opsLeft)
-        {
-            // pop from this one
-            filteredTxs.emplace_back(curTopTx);
-            cur->pop_front();
-            opsLeft -= opsCount;
-            // if there are more transactions, put it back
-            if (!cur->empty())
-            {
-                surgeQueue.push(cur);
-            }
-        }
-        else
-        {
-            // drop this transaction -> we need to drop the others
-            cur->clear();
-        }
+        auto perOpFee = computePerOpFee(*tx, lclHeader.ledgerVersion);
+        lowestFee = std::min(lowestFee, perOpFee);
     }
-    mTxs = TxSetUtils::sortTxsInHashOrder(filteredTxs);
+    mTxs = TxSetUtils::sortTxsInHashOrder(includedTxs);
```

Comment:
this conflicts with #3496 (so you'll have to rebase as we merge)

Reply:
Thanks for flagging this, I'll rebase after resolving this round of comments (for the sake of cleaner diffs)

Reply:
Rebased/merged this.